### PR TITLE
Inline fragments should not compile the front end template

### LINKED ISSUE #3263: Make sure Template#getResponse() is used correctly
**Affected version(s)**

Contao 4.9 and up

**Description**

see https://github.com/contao/contao/pull/3258#issuecomment-892492357

Currently, `Template#getResponse()` sets headers and executes `compile()` which will make the following code run (that is only meant to run once for `fe_page` or any other 'global wrapper' template):
https://github.com/contao/contao/blob/4cf71a9b3cf787093c9cf9e076d9c1fc6104f2f5/core-bundle/src/Resources/contao/classes/FrontendTemplate.php#L102-L159

- Should users in fragment controllers use `$template->getResponse()` or `new Response($template->parse())`?
- If so: Should we move things out to listeners?

### PR DESCRIPTION
**Pull request by aschempp**
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #3263

This is a draft attempt to not handle inline fragments like page content. However, ESI fragments **should** actually behave like the main page controller, it **should** minify markup etc.

An alternative idea by @fritzmg and me would be to register a response listener that minifies the main response content etc., but we're not sure that's a good idea 🙃

### CODE REVIEW

**Review by fritzmg — DISMISSED** [review]

File: core-bundle/src/Controller/AbstractFragmentController.php [hunk 1]
```diff
@@ -82,20 +83,27 @@ protected function getPageModel(): ?PageModel
...
-            $request = $this->get('request_stack')->getCurrentRequest();
-
             // Use the custom template unless it is a back end request
             if (null === $request || !$this->get('contao.routing.scope_matcher')->isBackendRequest($request)) {
                 $templateName = $model->customTpl;
             }
         }
 
-        $template = $this->get('contao.framework')->createInstance(FrontendTemplate::class, [$templateName]);
+        // Current request is the main request, which means this must be an ESI fragment
```
Comment (Toflar) [inline core-bundle/src/Controller/AbstractFragmentController.php]: Why ESI? Could also be hinclude or any other custom fragment strategy?

Comment (aschempp) [reply core-bundle/src/Controller/AbstractFragmentController.php]: yes it could be hinclude, but ESI is just the most use(d/ful) one 🤷

Comment (Toflar) [reply core-bundle/src/Controller/AbstractFragmentController.php]:
```suggestion
        // Current request is the main request, which means we must have a full template render, not just a fragment template (e.g. for ESI requests)
```

**Review by Toflar — DISMISSED** [review]

File: core-bundle/src/Resources/contao/classes/FragmentTemplate.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+		return $response;
+	}
+
+	protected function compile()
+	{
+		$this->strBuffer = $this->parse();
+	}
+}
+
+class_alias(FragmentTemplate::class, 'FragmentTemplate');
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/classes/FragmentTemplate.php]: This is not required for new classes.

**Review by leofeyer — COMMENTED** [review]
The initial comment says "This is a draft attempt" but the status of the PR is "ready for review". Which one is correct?

File: core-bundle/src/Resources/contao/classes/FragmentTemplate.php (hunk 2 above)
Comment (aschempp) [reply core-bundle/src/Resources/contao/classes/FragmentTemplate.php]: Removed in 1ec4e2d29ca8f41210d878e1216cd8a16b3b4115

**Comment by aschempp** [comment]
> The initial comment says "This is a draft attempt" but the status of the PR is "ready for review". Which one is correct?

Sorry about that. The draft was meant before we discussed this in the plenum at the dev meeting. Since we all agreed/couldn't think of a better way this is no longer considered a draft.

File: core-bundle/tests/Controller/ContentElement/ContentElementControllerTest.php [hunk 3]
```diff
@@ -243,9 +275,13 @@ private function mockContainerWithFrameworkTemplate(string $templateName): Conta
...
-        $container->set('contao.framework', $framework);
+        $this->container->set('contao.framework', $framework);
+
+        return $this->container;
+    }
 
-        return $container;
+    private function getContentModel(): ContentModel
+    {
+        return (new \ReflectionClass(ContentModel::class))->newInstanceWithoutConstructor();
```
Comment (leofeyer) [inline core-bundle/tests/Controller/ContentElement/ContentElementControllerTest.php:285]: Why is this necessary?

File: core-bundle/tests/Controller/ContentElement/ContentElementControllerTest.php [hunk 4]
```diff
@@ -228,11 +235,36 @@ public function testDoesNotSetTheSharedMaxAgeIfTheElementHasNeitherAStartNorASto
...
+        $currentRequest = new Request([], [], ['_scope' => 'frontend']);
+        $requestStack = $this->container->get('request_stack');
+
+        $requestStack->push(new Request()); // Main request
+        $requestStack->push($currentRequest); // Sub request
+
+        $controller = new TestController();
+        $controller->setContainer($this->container);
+
+        $controller(new Request(), $this->getContentModel(), 'main');
```
Comment (leofeyer) [inline core-bundle/tests/Controller/ContentElement/ContentElementControllerTest.php]: Should this not use `$currentRequest`?

```suggestion
        $controller($currentRequest, $this->getContentModel(), 'main');
```

File: core-bundle/tests/Controller/FrontendModule/FrontendModuleControllerTest.php [hunk 5]
```diff
@@ -158,10 +190,14 @@ private function mockContainerWithFrameworkTemplate(string $templateName): Conta
...
+        $this->container->set('contao.framework', $framework);
+        $this->container->set('contao.routing.scope_matcher', $this->mockScopeMatcher());
+
+        return $this->container;
+    }
 
-        return $container;
+    private function getModuleModel(): ModuleModel
+    {
+        return (new \ReflectionClass(ModuleModel::class))->newInstanceWithoutConstructor();
```
Comment (leofeyer) [inline core-bundle/tests/Controller/FrontendModule/FrontendModuleControllerTest.php]: Why can we no longer use `new ModuleModel()`?

Comment (aschempp) [reply core-bundle/tests/Controller/FrontendModule/FrontendModuleControllerTest.php]: Creating an instance of `Model` will extract DCA information and fail miserably without the framework booted.

File: core-bundle/tests/Controller/ContentElement/ContentElementControllerTest.php (hunk 4 above)
Comment (aschempp) [reply core-bundle/tests/Controller/ContentElement/ContentElementControllerTest.php]: Hah, yes it was meant to. But the test shows it works anyway, because the given one is not the main request 😅

**Review by leofeyer — APPROVED** [review]

**Comment by m-vo** [comment]
Returning a legacy class from a stub like so…

```php
        $framework = $this->mockContaoFramework();
        $framework
            ->expects($this->once())
            ->method('createInstance')
            ->with(FragmentTemplate::class, ['ce_test'])
            ->willReturn(new FragmentTemplate('ce_test'))
        ;
```

broke my local tests. The class was not autoloaded: `Error : Class "Contao\FragmentTemplate" not found`.

I think this is somewhat expected as our class loader isn't running, isn't it? I could fix it by requiring the actual file.


Edit: Hmm, I switched branches today and now the error isn't occuring anymore. :thinking: I do not have an explanation, though…
